### Optimizing A Data Set for Memory Usage

In [1]:
import pandas as pd

In [2]:
employees_data = '/Users/ypushiev/Learning/PANDAS IN ACTION/Chapter 5 Dataframe filtering/Data/employees.csv'

In [3]:
pd.read_csv(employees_data).isna().sum()

First Name     68
Gender        147
Start Date      2
Salary          2
Mgmt           68
Team           44
dtype: int64

In [4]:
df_employees = pd.read_csv(employees_data, parse_dates = ['Start Date'])

/var/folders/pg/x6kyk1l92fj8q4332g1j8rnw0000gp/T/ipykernel_35948/4257891297.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_employees = pd.read_csv(employees_data, parse_dates = ['Start Date'])


In [5]:
df_employees['Start Date'] = pd.to_datetime(df_employees['Start Date'], format='%Y-%m-%d')

In [6]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   First Name  933 non-null    object        
 1   Gender      854 non-null    object        
 2   Start Date  999 non-null    datetime64[ns]
 3   Salary      999 non-null    float64       
 4   Mgmt        933 non-null    object        
 5   Team        957 non-null    object        
dtypes: datetime64[ns](1), float64(1), object(4)
memory usage: 47.1+ KB


**memory usage: 47.1+ KB**

In [7]:
df_employees.head()

First Name  Gender Start Date    Salary   Mgmt       Team
0    Douglas    Male 1993-08-06       NaN   True  Marketing
1     Thomas    Male 1996-03-31   61933.0   True        NaN
2      Maria  Female        NaT  130590.0  False    Finance
3      Jerry     NaN 2005-03-04  138705.0   True    Finance
4      Larry    Male 1998-01-24  101004.0   True         IT

**Convert datetime into the object to cut the time from the data**

In [8]:
df_employees['Start Date'] = df_employees['Start Date'].dt.strftime('%Y-%m-%d')

In [9]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   First Name  933 non-null    object 
 1   Gender      854 non-null    object 
 2   Start Date  999 non-null    object 
 3   Salary      999 non-null    float64
 4   Mgmt        933 non-null    object 
 5   Team        957 non-null    object 
dtypes: float64(1), object(5)
memory usage: 47.1+ KB


**memory usage: 47.1+ KB**

### Converting Data Types with the astype Method

In [10]:
df_employees['Mgmt'].astype('bool').tail()

996     False
997     False
998     False
999      True
1000     True
Name: Mgmt, dtype: bool

**Convert object type into the bool type and see results in the size**

In [11]:
df_employees['Mgmt'] = df_employees['Mgmt'].astype('bool')

In [12]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   First Name  933 non-null    object 
 1   Gender      854 non-null    object 
 2   Start Date  999 non-null    object 
 3   Salary      999 non-null    float64
 4   Mgmt        1001 non-null   bool   
 5   Team        957 non-null    object 
dtypes: bool(1), float64(1), object(4)
memory usage: 40.2+ KB


**memory usage: 40.2+ KB**

### Convert into the integer type

**The Salary column contains NaN values**

In [13]:
df_employees['Salary'].isna().sum()

2

**Replace NaN values by 0**

In [14]:
df_employees['Salary']= df_employees['Salary'].fillna(0)

In [15]:
df_employees['Salary']=df_employees['Salary'].astype('int')

In [16]:
df_employees['Salary'].head()

0         0
1     61933
2    130590
3    138705
4    101004
Name: Salary, dtype: int64

In [17]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   First Name  933 non-null    object
 1   Gender      854 non-null    object
 2   Start Date  999 non-null    object
 3   Salary      1001 non-null   int64 
 4   Mgmt        1001 non-null   bool  
 5   Team        957 non-null    object
dtypes: bool(1), int64(1), object(4)
memory usage: 40.2+ KB


**The memory might be reduced more using INT32**

**Check max value**

In [18]:
df_employees['Salary'].max()

149908

In [19]:
df_employees['Salary']=df_employees['Salary'].astype('int32')

In [20]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   First Name  933 non-null    object
 1   Gender      854 non-null    object
 2   Start Date  999 non-null    object
 3   Salary      1001 non-null   int32 
 4   Mgmt        1001 non-null   bool  
 5   Team        957 non-null    object
dtypes: bool(1), int32(1), object(4)
memory usage: 36.3+ KB


In [21]:
df_employees['Salary'].max()

149908

In [22]:
df_employees['Salary'].head()

0         0
1     61933
2    130590
3    138705
4    101004
Name: Salary, dtype: int32

**memory usage: 36.3+ KB**

### Convert into categorical type

**Check unique values in the Dataframe** 

In [23]:
df_employees.nunique()

First Name    200
Gender          2
Start Date    971
Salary        995
Mgmt            2
Team           10
dtype: int64

**The Gender column contains only 2 unique values** 

In [24]:
df_employees['Gender'].astype('category').head()

0      Male
1      Male
2    Female
3       NaN
4      Male
Name: Gender, dtype: category
Categories (2, object): ['Female', 'Male']

In [25]:
df_employees['Gender'] = df_employees['Gender'].astype('category')

**Convert Team column type**

In [26]:
df_employees['Team'] = df_employees['Team'].astype('category')

In [27]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   First Name  933 non-null    object  
 1   Gender      854 non-null    category
 2   Start Date  999 non-null    object  
 3   Salary      1001 non-null   int32   
 4   Mgmt        1001 non-null   bool    
 5   Team        957 non-null    category
dtypes: bool(1), category(2), int32(1), object(2)
memory usage: 23.1+ KB


**memory usage: 23.1+ KB**

**The Dataframe size has been decreased from 47.1+ KB to 23.1+ KB approximately 50%**

### Filtering by a Single Condition

In [28]:
"Maria" == "Maria"

True

In [29]:
"Maria" == "Taylor"

False

**Check the column First Name**

In [30]:
df_employees["First Name"] == "Maria"

0       False
1       False
2        True
3       False
4       False
        ...  
996     False
997     False
998     False
999     False
1000    False
Name: First Name, Length: 1001, dtype: bool

**Select rows where "First Name" == "Maria"**

In [31]:
df_employees[df_employees["First Name"] == "Maria"]

First Name  Gender  Start Date  Salary   Mgmt          Team
2        Maria  Female         NaN  130590  False       Finance
198      Maria  Female  1990-12-27   36067   True       Product
815      Maria     NaN  1986-01-18  106562  False            HR
844      Maria     NaN  1985-06-19  148857  False         Legal
936      Maria  Female  2003-03-14   96250  False  Business Dev
984      Maria  Female  2011-10-15   43455  False   Engineering

**Create a filter for Dataframe**

In [32]:
marias = df_employees["First Name"] == "Maria"
df_employees[marias]

First Name  Gender  Start Date  Salary   Mgmt          Team
2        Maria  Female         NaN  130590  False       Finance
198      Maria  Female  1990-12-27   36067   True       Product
815      Maria     NaN  1986-01-18  106562  False            HR
844      Maria     NaN  1985-06-19  148857  False         Legal
936      Maria  Female  2003-03-14   96250  False  Business Dev
984      Maria  Female  2011-10-15   43455  False   Engineering

**Create a filter and find Top 5 Salary > 100000**

In [33]:
high_earners = df_employees["Salary"] > 100000
df_employees[high_earners].sort_values(by="Salary", ascending=False).head()

First Name  Gender  Start Date  Salary   Mgmt     Team
644  Katherine  Female  1996-08-13  149908  False  Finance
429       Rose  Female  2015-05-28  149903  False       HR
828    Cynthia  Female  2006-07-12  149684  False  Product
186        NaN  Female  2005-02-23  149654   True    Sales
160      Kathy  Female  2000-03-18  149563   True  Finance

### Filtering by Multiple Conditions

#### Condition and - &

**First condition**

In [34]:
is_female = df_employees['Gender'] == 'Female'

**Second condiotion**

In [35]:
in_biz_dev = df_employees['Team'] == "Business Dev"

**Find only women from the Business Development team**

In [36]:
df_employees[is_female & in_biz_dev].head()

First Name  Gender  Start Date  Salary   Mgmt          Team
9     Frances  Female  2002-08-08  139852   True  Business Dev
33       Jean  Female  1993-12-18  119082  False  Business Dev
36     Rachel  Female  2009-02-16  142032  False  Business Dev
38  Stephanie  Female  1986-09-13   36844   True  Business Dev
61     Denise  Female  2001-11-06  106862  False  Business Dev

In [37]:
is_manager = df_employees["Mgmt"]

**Add more conditionals**

In [38]:
df_employees[is_female & in_biz_dev & is_manager].head()

First Name  Gender  Start Date  Salary  Mgmt          Team
9      Frances  Female  2002-08-08  139852  True  Business Dev
38   Stephanie  Female  1986-09-13   36844  True  Business Dev
66       Nancy  Female  2012-12-15  125250  True  Business Dev
92       Linda  Female  2000-05-25  119009  True  Business Dev
111     Bonnie  Female  1999-12-17   42153  True  Business Dev

#### Condiotional OR - |

**Find employees with a salary lower than 40000 or a start date later than 2015-01-01**

In [39]:
earning_below_40k = df_employees["Salary"] < 40000
started_after_2015 = df_employees["Start Date"] > "2015-01-01"

In [41]:
df_employees[earning_below_40k | started_after_2015].tail()

First Name  Gender  Start Date  Salary   Mgmt         Team
958      Gloria  Female  1987-10-24   39833  False  Engineering
964       Bruce    Male  1980-05-07   35802   True        Sales
967      Thomas    Male  2016-03-12  105681  False  Engineering
989      Justin     NaN  1991-02-10   38344  False        Legal
1000        NaN     NaN         NaN       0   True          NaN

### Logical rejection (~)

#### Two same results with different ways of solving

**Salary lower than 100000 - strict conditional**

In [44]:
df_employees[df_employees["Salary"] < 100000].head()

First Name  Gender  Start Date  Salary  Mgmt         Team
0    Douglas    Male  1993-08-06       0  True    Marketing
1     Thomas    Male  1996-03-31   61933  True          NaN
6       Ruby  Female  1987-08-17   65476  True      Product
7        NaN  Female  2015-07-20   45906  True      Finance
8     Angela  Female  2005-11-22   95570  True  Engineering

**Using logical rejection**

In [45]:
df_employees[~(df_employees["Salary"] >= 100000)].head()

First Name  Gender  Start Date  Salary  Mgmt         Team
0    Douglas    Male  1993-08-06       0  True    Marketing
1     Thomas    Male  1996-03-31   61933  True          NaN
6       Ruby  Female  1987-08-17   65476  True      Product
7        NaN  Female  2015-07-20   45906  True      Finance
8     Angela  Female  2005-11-22   95570  True  Engineering

### Methods for boolean results

**Operation ==**

In [47]:
df_employees["Team"].eq("Marketing").head(1)

0    True
Name: Team, dtype: bool

**Operation !=**

In [48]:
df_employees["Team"].ne("Marketing").head(1)

0    False
Name: Team, dtype: bool

**Operation <**

In [50]:
df_employees["Salary"].lt(100000).head(1)

0    True
Name: Salary, dtype: bool

**Operation <=**

In [52]:
df_employees["Salary"].le(100000).head(1)

0    True
Name: Salary, dtype: bool

**Operation >**

In [51]:
df_employees["Salary"].gt(100000).head(1)

0    False
Name: Salary, dtype: bool

**Operation >=**

In [53]:
df_employees["Salary"].ge(100000).head(1)

0    False
Name: Salary, dtype: bool

### Filtering by Condition

#### The isin Method

In [56]:
all_star_teams = ["Sales", "Legal", "Marketing"]
is_in_all_star_teams = df_employees["Team"].isin(all_star_teams)
df_employees[is_in_all_star_teams].head()

First Name  Gender  Start Date  Salary   Mgmt       Team
0     Douglas    Male  1993-08-06       0   True  Marketing
5      Dennis    Male  1987-04-18  115163  False      Legal
11      Julie  Female  1997-10-26  102508   True      Legal
13       Gary    Male  2008-01-27  109831  False      Sales
20       Lois     NaN  1995-04-22   64714   True      Legal

#### The between Method

In [61]:
between_80k_and_90k = df_employees["Salary"].between(80000, 90000)
df_employees[['First Name','Team','Salary']][between_80k_and_90k].sort_values(by="Salary").head()

First Name          Team  Salary
821        NaN  Distribution   80399
865      Karen         Legal   80633
360      Susan         Sales   80688
104       John     Marketing   80740
542     Amanda  Distribution   80803

**Using for datetime type**

In [62]:
eighties_folk = df_employees["Start Date"].between(
left = "1980-01-01",
right = "1990-01-01"
)
df_employees[eighties_folk].head()

First Name  Gender  Start Date  Salary   Mgmt     Team
5      Dennis    Male  1987-04-18  115163  False    Legal
6        Ruby  Female  1987-08-17   65476   True  Product
10     Louise  Female  1980-08-12   63241   True      NaN
12    Brandon    Male  1980-12-01  112807   True       HR
17      Shawn    Male  1986-12-07  111737  False  Product

**Using for string type**

In [63]:
name_starts_with_r = df_employees["First Name"].between("R", "S")
df_employees[name_starts_with_r].head()

First Name  Gender  Start Date  Salary   Mgmt          Team
6        Ruby  Female  1987-08-17   65476   True       Product
36     Rachel  Female  2009-02-16  142032  False  Business Dev
45      Roger    Male  1980-04-17   88010   True         Sales
67     Rachel  Female  1999-08-16   51178   True       Finance
78      Robin  Female  1983-06-04  114797   True         Sales

### The isna and notna Methods

**isna method finds all NaN values from specific column**

In [66]:
no_team = df_employees["Team"].isna()
df_employees[['First Name','Gender','Team']][no_team].head()

First Name  Gender Team
1      Thomas    Male  NaN
10     Louise  Female  NaN
23        NaN    Male  NaN
32        NaN    Male  NaN
91      James     NaN  NaN

**notna method returns only True values in the "First Name" column**

In [69]:
has_name = df_employees["First Name"].notna()
df_employees[['First Name','Gender','Team']][has_name].head()

First Name  Gender       Team
0    Douglas    Male  Marketing
1     Thomas    Male        NaN
2      Maria  Female    Finance
3      Jerry     NaN    Finance
4      Larry    Male         IT

**Count only True values**

In [75]:
df_employees.notna().sum().sort_values(ascending=False)

Salary        1001
Mgmt          1001
Start Date     999
Team           957
First Name     933
Gender         854
dtype: int64

**Count only NaN values**

In [74]:
df_employees.isna().sum().sort_values(ascending=False)

Gender        147
First Name     68
Team           44
Start Date      2
Salary          0
Mgmt            0
dtype: int64

### Dealing with Null Values

In [77]:
df_dealna = pd.read_csv(employees_data, parse_dates = ['Start Date'])

/var/folders/pg/x6kyk1l92fj8q4332g1j8rnw0000gp/T/ipykernel_35948/883899616.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_dealna = pd.read_csv(employees_data, parse_dates = ['Start Date'])


In [79]:
df_dealna.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   First Name  933 non-null    object        
 1   Gender      854 non-null    object        
 2   Start Date  999 non-null    datetime64[ns]
 3   Salary      999 non-null    float64       
 4   Mgmt        933 non-null    object        
 5   Team        957 non-null    object        
dtypes: datetime64[ns](1), float64(1), object(4)
memory usage: 47.1+ KB


#### dropna method will remove all rows where NaN values exist

In [81]:
df_dealna.dropna().isnull().sum()

First Name    0
Gender        0
Start Date    0
Salary        0
Mgmt          0
Team          0
dtype: int64

**Dropna(how = 'all') removes only rows where all values are NaN**

In [84]:
df_dealna.shape


(1001, 6)

In [85]:
drop_how = df_dealna.dropna(how='all')
drop_how.shape

(1000, 6)

**Drop NaN values in the specific column**

In [89]:
df_dealna.dropna(subset=['Gender']).head()

First Name  Gender Start Date    Salary   Mgmt       Team
0    Douglas    Male 1993-08-06       NaN   True  Marketing
1     Thomas    Male 1996-03-31   61933.0   True        NaN
2      Maria  Female        NaT  130590.0  False    Finance
4      Larry    Male 1998-01-24  101004.0   True         IT
5     Dennis    Male 1987-04-18  115163.0  False      Legal

**Cleaning a few columns**

In [91]:
df_dealna.dropna(subset=['Team','Salary','Gender']).isnull().sum()

First Name    52
Gender         0
Start Date     1
Salary         0
Mgmt          52
Team           0
dtype: int64

**The thresh parameter allows you to set a minimum number of non-empty values
in line**

In [ ]:
# Returns all True values because of thresh=6 and Dataframe has only 6 columns
df_dealna.dropna(thresh=6).head()

First Name  Gender Start Date    Salary   Mgmt          Team
4      Larry    Male 1998-01-24  101004.0   True            IT
5     Dennis    Male 1987-04-18  115163.0  False         Legal
6       Ruby  Female 1987-08-17   65476.0   True       Product
8     Angela  Female 2005-11-22   95570.0   True   Engineering
9    Frances  Female 2002-08-08  139852.0   True  Business Dev

In [ ]:
# returns all rows with at least 4 non-NA values
df_dealna.dropna(thresh=4).head()

First Name  Gender Start Date    Salary   Mgmt       Team
0    Douglas    Male 1993-08-06       NaN   True  Marketing
1     Thomas    Male 1996-03-31   61933.0   True        NaN
2      Maria  Female        NaT  130590.0  False    Finance
3      Jerry     NaN 2005-03-04  138705.0   True    Finance
4      Larry    Male 1998-01-24  101004.0   True         IT